### PR TITLE
[Security] Add concept of required passport badges

### LINKED ISSUE #39305: [Security] Mark badges as required
[Just a placeholder, to make sure I don't forget, I'll add details later]

During the SymfonyWorld keynote, people suggested to add an option like `required_badges` were you can *on an application level* require which badges should be set on the password. This should avoid e.g. loading a 3rd party authenticator that somehow removed the `CsrfBadge`  badge - while you think you still have CSRF support.

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | n
| Tickets       | Fix #39305
| License       | MIT
| Doc PR        | tbd

A badge on a passport is a critical security element, it determines which security checks are run during authentication. Using the `required_badges` setting, applications can make sure the expected security checks are run.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 1]
```diff
@@ -266,6 +268,27 @@ private function addFirewallsSection(ArrayNodeDefinition $rootNode, array $facto
...
+                ->info('A list of badges that must be present on the authenticated passport.')
+                ->validate()
+                    ->always()
+                    ->then(function ($requiredBadges) {
+                        return array_map(function ($requiredBadge) {
+                            if (class_exists($requiredBadge)) {
+                                return $requiredBadge;
+                            }
+
+                            $fqcn = 'Symfony\Component\Security\Http\Authenticator\Passport\Badge\\'.$requiredBadge;
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: should we attempt to prepend the symfony security namespace of core badges if there is already a `\` in the string ?

**Review by stof — COMMENTED** [review]
Don't we need to update the XSD schema for the XML format ?

**Comment by wouterj** [comment]
> Don't we need to update the XSD schema for the XML format ?

Yes, but we need to do it in 5.2: https://github.com/symfony/symfony/pull/40490 (or 5.1 actually, but that's no longer supported)

**Comment by stof** [comment]
@wouterj you will still need an update in this PR to add the `required-badge` node in the XSD

File: src/Symfony/Component/Security/Http/EventListener/CheckRequiredBadgesListener.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        if ($missing) {
+            throw new BadCredentialsException(sprintf('Authentication failed; Some badges marked as required by the firewall config are not available on the passport: "%s".', implode('", "', $missing)));
+        }
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CheckPassportEvent::class => ['checkPassport', -512],
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Http/EventListener/CheckRequiredBadgesListener.php]: does -512 imply we expect new badges during the checking stage?

shouldnt passport be immutable in that regard?

Comment (wouterj) [reply src/Symfony/Component/Security/Http/EventListener/CheckRequiredBadgesListener.php]: They aren't, e.g. `UserProviderListener` automatically adds the password upgrade badge if your user provider supports upgrading passwords (and there are password credentials).

**Review by Nyholm — APPROVED** [review]
Thank you. 

Im happy with this. 

I have a question though.. What if I register another listener to the `CheckPassportEvent` and it will run `stopPropagation()`? Now this event subscriber will never be executed and I wont get any exception. Right?

It is just an observation, it might be out of scope for this PR but I would still ask if there is something we can/should do about this scenario.

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 5.3
 ---
 
+ * Add `CheckRequiredBadgesListener` to validate which badges are present
```
Comment (chalasr) [inline src/Symfony/Component/Security/CHANGELOG.md]: no longer true :)

**Comment by wouterj** [comment]
> It is just an observation, it might be out of scope for this PR but I would still ask if there is something we can/should do about this scenario.

Thanks for reminding me :) After #40567, we can do this in a much "cleaner" way. PR updated

**Review by chalasr — APPROVED** [review]
LGTM. The new config option should probably be mentioned in the SecurityBundle' CHANGELOG file though

**Comment by Nyholm** [comment]
Thank you @wouterj.

**Pull request merged**